File: Edendale.Windows/Models/UserMediaModels.cs
// Windows user-media and TMDB account models.

namespace Edendale.Windows.Models;

/// <summary>
/// Per-title user state keyed by TMDB id + media type ("movie"/"tv").
/// Each field carries its own last-write timestamp (cloud-replica merge) and
/// dirty flag (pending TMDB account push).
/// </summary>
public sealed class UserMediaRecord
{
    public int TmdbId { get; set; }
    public string MediaType { get; set; } = "movie";
    public string? Title { get; set; }
    public string? PosterPath { get; set; }
    public string? PosterUrl { get; set; }
    public bool Favourite { get; set; }
    public long FavouriteUpdatedAt { get; set; }
    public bool FavouriteDirty { get; set; }
    public bool Watchlist { get; set; }
    public long WatchlistUpdatedAt { get; set; }
    public bool WatchlistDirty { get; set; }
    public double? Rating { get; set; }
    public long RatingUpdatedAt { get; set; }
    public bool RatingDirty { get; set; }

    public string StorageKey => $"{MediaType}:{TmdbId}";

    /// <summary>Stateless records are pruned from local and remote replicas.</summary>
    public bool HasState =>
        Favourite || Watchlist || Rating is not null ||
        FavouriteDirty || WatchlistDirty || RatingDirty;
}

/// <summary>Step one of the TMDB connect flow.</summary>
public sealed class TmdbAuthStart
{
    public string RequestToken { get; set; } = "";
    public string ApprovalUrl { get; set; } = "";
}

/// <summary>A signed-in TMDB session; persisted DPAPI-protected on disk.</summary>
public sealed class TmdbSessionInfo
{
    public string SessionId { get; set; } = "";
    public int AccountId { get; set; }
    public string? Username { get; set; }
    public string? Name { get; set; }
}

/// <summary>Result of one TMDB sync round; Records replaces the local store.</summary>
public sealed class UserMediaSyncOutcome
{
    public List<UserMediaRecord> Records { get; set; } = [];
    public int Pushed { get; set; }
    public int Pulled { get; set; }
}

File: src/main/kotlin/com/babasama/edendale/domain/UserMediaModels.kt
package com.babasama.edendale.domain

/** TMDB's rating domain: 0.5–10 in half-point steps; null means unrated. */
fun sanitizeTmdbRating(value: Double?): Double? = value
    ?.let { (it * 2).toInt() / 2.0 }
    ?.coerceIn(0.5, 10.0)

/**
 * Per-title user state (favourite / watchlist / rating) keyed by TMDB id +
 * media type. Each field carries its own last-write timestamp so replicas on
 * other devices merge field-wise (newest wins), and a dirty flag so changes
 * made while disconnected are pushed on the next TMDB account sync.
 *
 * The optional display snapshot (title/poster) lets watchlist and favourite
 * shelves render without a network fetch per stored id.
 */
data class UserMediaRecord(
    val tmdbId: Int,
    val mediaType: MediaType,
    val title: String? = null,
    val posterPath: String? = null,
    val favourite: Boolean = false,
    val favouriteUpdatedAt: Long = 0L,
    val favouriteDirty: Boolean = false,
    val watchlist: Boolean = false,
    val watchlistUpdatedAt: Long = 0L,
    val watchlistDirty: Boolean = false,
    val rating: Double? = null,
    val ratingUpdatedAt: Long = 0L,
    val ratingDirty: Boolean = false,
) {
    val ref: MediaRef get() = MediaRef(tmdbId, mediaType)
    val storageKey: String get() = "${mediaType.pathSegment}:$tmdbId"

    /**
     * True while the record still says something: a set flag or rating, or a
     * pending un-set that has not been pushed to TMDB yet. Records without
     * state are pruned rather than stored.
     */
    val hasState: Boolean
        get() = favourite || watchlist || rating != null ||
            favouriteDirty || watchlistDirty || ratingDirty

    fun settingFavourite(value: Boolean, nowMillis: Long): UserMediaRecord =
        copy(favourite = value, favouriteUpdatedAt = nowMillis, favouriteDirty = true)

    fun settingWatchlist(value: Boolean, nowMillis: Long): UserMediaRecord =
        copy(watchlist = value, watchlistUpdatedAt = nowMillis, watchlistDirty = true)

    fun settingRating(value: Double?, nowMillis: Long): UserMediaRecord =
        copy(rating = sanitizeTmdbRating(value), ratingUpdatedAt = nowMillis, ratingDirty = true)

    /** Fills the display snapshot without disturbing state or timestamps. */
    fun withDisplay(title: String?, posterPath: String?): UserMediaRecord = copy(
        title = title ?: this.title,
        posterPath = posterPath ?: this.posterPath,
    )
}
